urbs.py

```diff
@@ -583,13 +633,72 @@ def res_vertex_rule(m, tm, sit, com, com_type):
     # if com is a demand commodity, the power_surplus is reduced by the
     # demand value; no scaling by m.dt or m.weight is needed here, as this
     # constraint is about power (MW), not energy (MWh)
+    # added demand side management
     if com in m.com_demand:
         try:
-            power_surplus -= m.demand.loc[tm][sit, com]
+                if tm <= m.timesteps[0] + m.L:
+                    power_surplus -= m.demand.loc[tm][sit, com] \
+				+ m.DSMup[tm,sit] - sum(m.DSMdo[T,tm,sit] for T in range(m.timesteps[0] + 1, tm+m.L+1))
+                elif tm >= m.timesteps[0] + 1 + m.L and tm <= m.timesteps[-1] - m.L:
+                    power_surplus -= m.demand.loc[tm][sit, com] \
+                       + m.DSMup[tm,sit] - sum(m.DSMdo[T,tm,sit] for T in range(tm-m.L, tm+1+m.L))
+                else:
+                    power_surplus -= m.demand.loc[tm][sit, com] \
+				+ m.DSMup[tm,sit] - sum(m.DSMdo[T,tm,sit] for T in range(tm-m.L, m.timesteps[-1] + 1))   
```

Comment:
This looks like it could need its own helper function, much like `commodity_balance`.
Possible problem: the code assumes that timestep labels are
 * numeric (1, 4, 7, 2401, 8760) and
 * consecutive (5, 6, 7, 8, 9, 10).
However, neither assumption must hold true. Time step labels could become strings, datetime objects, ...